src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+    }
+
+    private void unlink(Node<T> node) {
+        Node<T> next = node.next;
+        Node<T> prev = node.prev;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+
+        node.item = null;
+        size--;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> current = getNodeAt(index);
+        Node<T> newNode = new Node<>(current.prev, value, current);
+
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        current.prev = newNode;
+        size++;
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeAt(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToUpdate = getNodeAt(index);
+        T oldValue = nodeToUpdate.item;
+        nodeToUpdate.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAt(index);
+        T removedValue = current.item;
+
+        if (current.prev != null) {
+            current.prev.next = current.next;
+        } else {
+            head = current.next;
+        }
+
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        } else {
+            tail = current.prev;
+        }
+
+        current.item = null;
+        size--;
```

Comment:
The `remove(int index)` method duplicates some logic found in the `unlink(Node<T> node)` method. Consider using the `unlink` method to avoid code duplication and improve maintainability.